pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaTableReadFlow.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.pravega.impl
+
+import java.util.concurrent.CompletableFuture
+
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.stream.stage.{AsyncCallback, GraphStage, GraphStageLogic, InHandler, OutHandler, StageLogging}
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+
+import scala.util.control.NonFatal
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import akka.stream.alpakka.pravega.TableSettings
+
+import scala.util.{Failure, Try}
+import io.pravega.client.tables.KeyValueTable
+import io.pravega.client.KeyValueTableFactory
+
+import io.pravega.client.tables.KeyValueTableClientConfiguration
+
+import java.util.concurrent.atomic.AtomicInteger
+import io.pravega.client.tables.TableEntry
+import scala.util.Success
+@InternalApi private final class PravegaTableReadFlowStageLogic[K, V](
+    val shape: FlowShape[K, Option[V]],
+    val scope: String,
+    tableName: String,
+    tableSettings: TableSettings[K, V],
+    familyExtractor: K => String
+) extends GraphStageLogic(shape)
+    with StageLogging {
+
+  private def in = shape.in
+  private def out = shape.out
+
+  private var keyValueTableFactory: KeyValueTableFactory = _
+  private var table: KeyValueTable[K, V] = _
+
+  private val onAir = new AtomicInteger
```

Comment:
This represents the number of messages in-flight?

Reply:
As long as it's being mutated during standard stream stage events this can just be a `@volatile var`. Async callbacks execute in serial with other stage events.